matrixprofile/algorithms/streaming.py

```diff
@@ -0,0 +1,398 @@
+            # This can arise when aggregating multiple sequences of non-uniform length.
+            # One sequence may depend on > 1 elements of another sequence, so a shift by some amount
+            # greater than what is contained here right now indicates no data was provided. For that reason
+            # we apply the entire shift to min_index, implicitly normalize the buffer, and set the number of live
+            # elements to 0 (so array.data returns an empty view)
+
+            self.begin_pos = 0
+            self.count = 0
+        else:
+            self.begin_pos += count
+            self.count -= count
+        self.min_index += count
+
+    def resize_buffer(self, size):
+        """ resize underlying buffer, raise an error if it would truncate live data """
+        if size < self.count:
+            raise ValueError(f"buffer size {size} is too small to accommodate {self.count} live elements")
+        data = self.array
+        self._array = np.empty(size, dtype=self._array.dtype)
+        self.array[:] = data
+
+    def append(self, data):
+        """ append to buffer. In most python interfaces, extend is used with iterables. This naming convention
+            maintains consistency with that, without checking explicit inheritance from Iterable, which numpy may
+            not adhere to strictly.
+        """
+        reqspace = data.size
+        if self.free_count < reqspace:
+            raise ValueError(f"appending {dat.size} elements would overflow available buffer space: {self.free_count}")
+        prevct = self.count
+        self.count += reqspace
+        self.array[prevct:] = data
+
+    def normalize_buffer(self):
+        """ normalize buffer layout so that retained data spans the positional range
+            0 to size - 1.
+            Since this is an inplace op, it can invalidate python iterators to this object.
+            It is intentionally left as an explicit op.
+        """
+        if self.begin_pos != 0:
+            self._array[:self.count] = self.array
+            self.begin_pos = 0
+
+
+def xcov(ts, mu, cmpto, out=None):
+    sseqct = ts.shape[0] - cmpto.shape[0] + 1
+    if sseqct > 0 and out is None:
+        out = np.empty(sseqct - minsep, dtype='d')
+    mps.crosscov(out, ts, mu, cmpto)
+    return out
+
+
+def mpx(ts, w):
+    if w < 2:
+        raise ValueError(f"Window length must be at least 2, received {w}")
+    sseqct = ts.shape[0] - w + 1
+    mu = np.empty(sseqct, dtype='d')
+    mu_s = np.empty(sseqct, dtype='d')  # window length w - 1 skipping first and last
+    invn = np.empty(sseqct, dtype='d')
+    minsep = w // 4
+    if sseqct <= minsep:
+        return np.full(sseqct, -1.0), np.full(sseqct, -1)
+    mps.windowed_mean(ts, mu, w)
+    mps.windowed_mean(ts[:-1], mu_s, w - 1)
+    mps.windowed_invcnorm(ts, mu, invn, w)
+    rbwd = ts[:sseqct - 1] - mu[:sseqct - 1]
+    cbwd = ts[:sseqct - 1] - mu_s[1:]
+    rfwd = ts[w:] - mu[1:]
+    cfwd = ts[w:] - mu_s[1:]
```

Comment:
It should not be encapsulated in cases where it might be reused. My preferred solution would have been to use the functions in cympx, but they are currently too monolithic, so I threw this together.